### PR TITLE
Wp pycanon

### PR DESCRIPTION
**Pull request by willpoulett**
# Summary

**What/Why**

Added pycanon and changed some config settings so you can turn on/off pycorrectmatch and shap. Added documentation.

**Your approach**

allows you to turn off sections of the pipeline if code doesn't work.
**What to review**

Include any helpful pointers for the reviewer.

# Checklists

This pull/merge request meets the following requirements:

- [x] code runs
- [x] [developments are ethical][data-ethics-framework]
- [x] developments are secure (no data or notebook output)
- [x] you have made proportionate checks that the code works correctly
- [x] [minimum usable documentation][agilemodeling] written in the `docs` folder

Comments have been added below around the incomplete checks.

[agilemodeling]: http://agilemodeling.com/essays/documentLate.htm
[data-ethics-framework]: https://www.gov.uk/government/publications/data-ethics-framework

### CODE REVIEW

File: docs/pycanon/pycanon_and_privacy_metrics.md [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+As defined in the PyCanon [paper](https://www.nature.com/articles/s41597-022-01894-2), a database verifies K-anonymity if for each row of the database, there are at least k-1 indistinguishable rows with respect to the quasi-identifiers. The value of K deemed acceptable has to be decided beforehand. The probability of identifying an individual in the database using the quasi-identifiers will be at most 1/k.
+
+### (Alpha,K)-Anonymity
+
+Given only one sensitive attribute S, it is checked if the database is k-anonymous and the frequency of each possible value of S is lower or equal than α in each equivalence class.
+
+### L-Diversity
+
+K-anonymity gets you equivalence classes - sets of records in the dataset which share the same values for their quasi identifiers. L-diversity is an extension of this. In the case of one sensitive attribute, it checks whether there are at least L distinct values for S. L must always be greater than or equal to 1.
```
[the remark this answers is not in the document]
Comment (danjscho) [reply docs/pycanon/pycanon_and_privacy_metrics.md:25]: Same throughout the below

File: docs/pycanon/pycanon_and_privacy_metrics.md [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+Following this idea, we instead can explore the use of anomoly detection to identify high risk columns.
+
+### Distance Based Methods
+
+Distance metrics can be used to find outliers. The greater the distance away from other individuals, the more unique they are. Possible [distance metrics](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.DistanceMetric.html) that can be used are:
+
+- Euclidean Distance
+- Manhatten Distance
+- Minowski Distance
```
Comment (danjscho) [inline docs/pycanon/pycanon_and_privacy_metrics.md:92]: Add any thoughts on where best to use each one?

File: docs/pycanon/pycanon_and_privacy_metrics.md [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+**Identifier:** This is any piece of information, such as a full name or NHS number, which can directly identify an individual.
+
+**Quasi-Identifier** This is a piece of information that when combined with other Quasi-Identifiers can make identification possible. This could include birthdays and education information.
+
+**Sensitive Attributes:** Sensitive attributes contain private information that must not be extracted.
+
+### K-Anonymity
+
+As defined in the PyCanon [paper](https://www.nature.com/articles/s41597-022-01894-2), a database verifies K-anonymity if for each row of the database, there are at least k-1 indistinguishable rows with respect to the quasi-identifiers. The value of K deemed acceptable has to be decided beforehand. The probability of identifying an individual in the database using the quasi-identifiers will be at most 1/k.
```
[the remark this answers is not in the document]
Comment (willpoulett) [reply docs/pycanon/pycanon_and_privacy_metrics.md:17]: A quick look online seems that people use both capital or not. I will go for no caps.

File: docs/pycanon/pycanon_and_privacy_metrics.md (hunk 2 above)
Comment (willpoulett) [reply docs/pycanon/pycanon_and_privacy_metrics.md:92]: Added a couple of sentences:

Different distance metrics will have an effect on which data points are considered outliers. As our data is unlikely to be high-dimensional, simple methods such as Euclidean Distance can be both effective and explainable. However, for higher dimensional multi-variate data, more advanced distance metrics could be used, such as Mahalanobis which is effective for high dimensional multi-variate data.

File: config/experimental_config.yaml [hunk 4]
```diff
@@ -28,3 +28,18 @@ extraction:
...
+  shap: true
+
+pycanon:
+  identifiers:
+    - nhs number
+  quasi_identifiers:
+    - date of birth
+    - person
+  sensitive_attributes:
+    - diagnosis
```
[the remark this answers is not in the document]
Comment (willpoulett) [reply config/experimental_config.yaml:45]: Added a piece of markdown on experimental pipeline notebook explaining this.

File: notebooks/full_pipeline_example.ipynb [hunk 5]
```diff
@@ -495,11 +514,59 @@
...
+    "sensitive_attributes = experimental_config.pycanon.sensitive_attributes"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from pycanon.anonymity import k_anonymity, t_closeness, l_diversity\n",
```
Comment (danjscho) [inline notebooks/full_pipeline_example.ipynb:553]: Move import to top of notebook

File: requirements.txt [hunk 6]
```diff
@@ -38,6 +38,7 @@ omegaconf==2.3.0
 plotly==5.22.0
 pre-commit==3.6.1
 prompttools==0.0.43
+pycanon[PDF]==1.0.1
 pydantic==2.7.1
 scipy==1.12
 sentencepiece==0.2.0
```
Comment (danjscho) [inline requirements.txt:44]: Could you put the full file in alphabetical order please :-)

**Review by danjscho — CHANGES_REQUESTED** [review]
Two small things, then good to go

**Pull request merged**